### PR TITLE
Tconstr syntax

### PR DESCRIPTION
**Pull request by mlutze**
Fixes https://github.com/flix/flix/issues/1492

### Features
* Can use `with` postfix syntax for defs and sigs
  * [x] impl
  * [x] test
* Tparam kinds are annotated with `:`
  * [x] impl
  * [x] test
* Tparam classes are annotated with `has`
  * [x] impl
  * [x] test
 
* [x] Decide what to do with multiple constraints on the same tparam.

### General requirements
* [x] Don't make a complete mess
* [x] Don't break anything else
* [x] Provide documentation
* [x] Add license to new files

### CODE REVIEW

**Comment by mlutze** [comment]
@magnus-madsen RFR

Note that currently for multiple constraints on a tparam we just repeat the word "has" (since this made automated changes easier).
E.g. `def f[a has Functor has ToString](x: a): a = ???`

Ugly but pending better decision

**Comment by magnus-madsen** [comment]
Pending outcome of #1492 right?

**Comment by mlutze** [comment]
> Pending outcome of #1492 right?

Right

**Comment by mlutze** [comment]
@magnus-madsen Now really ready for review :)

**Comment by magnus-madsen** [comment]
And its a big one! :)

File: main/test/ca/uwaterloo/flix/library/TestToString.flix [hunk 1]
```diff
@@ -375,4 +375,8 @@ namespace TestToString {
     def arrayToString05(): Bool & Impure =
         toString(Array.new([1, 2, 3], 3)) == "[[1, 2, 3], [1, 2, 3], [1, 2, 3]]"
 
+    @test
+    def arrayToString06(): Bool & Impure =
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestToString.flix:379]: 👍

File: main/src/library/GetOpt.flix [hunk 2]
```diff
@@ -1,418 +1,418 @@
-/*
```
Comment (magnus-madsen) [inline main/src/library/GetOpt.flix:1]: Why does this entire file show up as changed? What changed??

**Comment by magnus-madsen** [comment]
This PR allows:

```
pub def memberOf[a](x: a, a: Array[a]): Bool & Impure with Eq[a] =
```

so the plan is to revisit that later and demand that `a` be given a kind?

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 3]
```diff
@@ -242,9 +238,13 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
       SP ~ Names.QualifiedClass ~ optWS ~ "[" ~ optWS ~ Type ~ optWS ~ "]" ~ SP ~> ParsedAst.TypeConstraint
     }
 
+    def OptTypeConstraintList: Rule1[Seq[ParsedAst.TypeConstraint]] = rule {
+      optional(optWS ~ keyword("with") ~ optWS ~ oneOrMore(TypeConstraint).separatedBy(optWS ~ "," ~ optWS)) ~> ((o: Option[Seq[ParsedAst.TypeConstraint]]) => o.getOrElse(Seq.empty))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:242]: Should it really be optWS here? Would that not allow you to write e.g. `withEq[a]`??

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:242]: `keyword` has a lookahead that prevents this. But could change to `WS` for consistency.

File: main/src/library/GetOpt.flix (hunk 2 above)
Comment (mlutze) [reply main/src/library/GetOpt.flix:1]: The file mode changed (100755 → 100644). I guess it was executable before?

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala (hunk 3 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:242]: But what about the `optWS` before? I think we should put hard WS in both places.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:242]: Can you explore what happened? If you cannot find anything then we can just merge it.

File: main/src/library/GetOpt.flix (hunk 2 above)
Comment (mlutze) [reply main/src/library/GetOpt.flix:1]: Looks like this file was CRLF. When I piped everything through a script, it changed it to LF.

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala (hunk 3 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:242]: I don't know if we should be so strict when there's no ambiguity.

Scala allows e.g. `class C[a]extends T` (no space before `extends`)

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:242]: Ah- but we don't allow such things anywhere currently.

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala]: Fair enough. Fixed.

**Comment by magnus-madsen** [comment]
Excellent!

**Pull request merged**